Add current sprint process, GH projects-world, and experiments Inframundo is running

jdangerx opened the pull request:
### Describe the Policy Change

After discussion in https://github.com/catalyst-cooperative/pudl/pull/2304 we decided to:

* move the documentation to the handbook
* write down our existing process
* write down how to port that to Github projects (which turns out to be... basically the same)
* explicitly call out experiments we'd like to run

Here is all that writing!

e-belfer commented on docs/catalyst_101.md:
```diff
@@ -136,6 +136,153 @@ Once the existing members have voted to offer you membership, it’s up to you t
+- We should be able to identify dependencies and bottlenecks.
+- Each team should be able to communicate its goals on a two-week timeline to the other teams.
+- We should be able to stick to the process, because the overhead feels worth the effort.
+
```
We should be able to break down our quarterly goals into smaller, logical steps and work towards them consistently over the quarter.

jdangerx replied:
This sounds like two things:
* we would like something between quarterly and sprint planning
* in the sprint we should be connecting the work we do to progress on our high level goals

I put some stuff in to that effect - let me know if I'm misinterpreting your words!

e-belfer replied:
Exactly it!

e-belfer commented on docs/catalyst_101.md:
```diff
@@ -136,6 +136,153 @@ Once the existing members have voted to offer you membership, it’s up to you t
+
+In the sprint review, instead of discussing status for each item in the sprint, prioritize the tickets that are in Backlog according to the following schema:
+
+- "Urgent" -> blocking work on high-priority tickets
```
or affecting overall functioning of PUDL for users?

jdangerx replied:
I like having user-focused things here, but I'm a little wary of this specific language because it's very broad.

I'd like for Urgent to be a pretty clearly and tightly defined thing - otherwise lots of stuff goes into Urgent and then we have no prioritization scheme anymore 😢 

I think a few specific user-facing services that we provide being labeled "urgent" would be totally reasonable, though! Maybe worth cataloging exactly what sorts of stuff we do for people 😄

e-belfer commented on docs/catalyst_101.md:
```diff
@@ -136,6 +136,153 @@ Once the existing members have voted to offer you membership, it’s up to you t
+
+Sprint planning (Monday, week 1):
+
+- for in-flight items in the old sprint:
```
What do you mean by in-flight? As in, someone is actively working on it but it's not done?

jdangerx replied:
Yep - In progress & review. Let me update the docs.

zschira commented on docs/catalyst_101.md:
```diff
@@ -136,6 +136,159 @@ Once the existing members have voted to offer you membership, it’s up to you t
+
+Pretty much the same, but with the Catalyst Backlog project instead of ZenHub.
+
+Sprint planning (Monday, week 1):
```
Maybe including some high level discussion about what we'd like to have done this sprint, and how it relates to bigger goals will help us with setting priorities

zaneselvans replied:
I like having these higher level functional goals at the sprint level too, like "Have the ETL migrated to Dagster." even though it may be composed of several issues.

On the plus side, I think it can help with prioritization of issues -- if we decide the sprint has a core purpose, it makes it easier to set aside peripheral things that are going on, or interruptions that might come up.

On the minus side, I think we sometimes set overly ambitious high level goals because it's nice to hit those bigger milestones, and then we don't complete them by the end of the sprint, and that feels bad / demotivating.

jdangerx replied:
A lot of people like having high level goals for sprints - that also helps us prioritize within the sprint. We sort of did that ad-hoc with this most recent one by saying "get dagster merged is the top priority" and then prioritizing blockers to that as "high" or "urgent" and everything else as "medium" or "low".

That would probably go in the "experiments" section though, not the "100% faithful port of our existing process from Zenhub to GH Projects" section.

zschira commented on docs/catalyst_101.md:
```diff
@@ -136,6 +136,159 @@ Once the existing members have voted to offer you membership, it’s up to you t
+
+- Say no to new things if they would push you over WIP limits.
+
+Success criteria:
```
Maybe another success criteria could be something like: "- does not block important work from getting done". I think it's important to make sure that limits we're setting are not unnecessarily restrictive such that it blocks us from being able to move on to new work when there's not really a bottleneck.

jdangerx replied:
I love it, adding!

zschira commented on docs/catalyst_101.md:
```diff
@@ -136,6 +136,159 @@ Once the existing members have voted to offer you membership, it’s up to you t
+
+Accountability/how we'll stick to it:
+
+- Just the in-meeting Thing Counting Phase. Maybe we can make it a race to make it engaging.
```
Is there any kind of github automation we could do to enforce this automatically? And if so, would implementing that automation be overly complicated?

jdangerx replied:
I think we could do some sort of script that cleans things up via the API but I don't think we can enforce a WIP limit within the UI.

zschira commented on docs/catalyst_101.md:
```diff
@@ -136,6 +136,159 @@ Once the existing members have voted to offer you membership, it’s up to you t
+
+Sprint review (Thursday, week 2):
+
+- for each item in the sprint:
```
Similar to previous comment, but I like the idea of having some discussion in each of these meetings about how progress relates to higher level goals

jdangerx replied:
Yeah, I think this is a good idea! Our meetings might start to get kind of long if we're adding backlog tending as an actual activity that we do - I kind of like separating backlog tending & sprint retro from each other, but we can see how shoving everything into sprint review feels for now.

jdangerx replied:
I guess "have high level goals for the sprint" would be a new experiment that captures both of these.

zschira reviewed (commented):
Excited to keep experimenting with these things, thanks @jdangerx!

zaneselvans commented on docs/catalyst_101.md:
```diff
@@ -136,6 +136,159 @@ Once the existing members have voted to offer you membership, it’s up to you t
+  - discuss whether we should pull into the new sprint; tag with new sprint if necessary.
+- ad-hoc discussion topics
+
+Mid sprint checkin (Monday, week 2):
```
Do we want to include any kind of descoping / rescoping into the mid-sprint check-in? Or do we want to treat the sprint contents as unalterable between the beginning and end (which seems like it might be more helpful for learning to scope sprints appropriately)

jdangerx replied:
I think mid-sprint checkin is a great place for people to flag "this ticket is blowing up, we need to re-scope." I think forcing ourselves to not talk about scope because we should have scoped better in the first place is counter-productive.

zaneselvans commented on docs/catalyst_101.md:
```diff
@@ -136,6 +136,159 @@ Once the existing members have voted to offer you membership, it’s up to you t
+
+- for each item in the sprint:
+  - discuss status
+- ad-hoc discussion topics
```
I'm a little apprehensive about including ad-hoc discussion in every one of these meetings, especially if there are more than ~3 people. In the past we've definitely had the sprint planning / review meetings wander off into longer unstructured discussions maybe inadvisably.

jdangerx replied:
Yeah, I don't like it either, but this section is the "here is what we are doing now" section and I didn't want to elide this part of the meeting.

zaneselvans commented on docs/catalyst_101.md:
```diff
@@ -136,6 +136,159 @@ Once the existing members have voted to offer you membership, it’s up to you t
+
+Outside of the meetings:
+
+- If a new issue comes up, it goes in the Icebox unless there's a compelling case for it to be Urgent, in which case we talk about it as a team and pull it into the sprint if it's truly Urgent.
```
How do issues get out of the Icebox? If all New issues go there by default unless they're Urgent will we ever see them again?

I've imagined having a bi-weekly process for categorizing New issues -- deciding either that they're something we do want to tackle and fleshing them out enough to go into the Project Backlog, or that they're beyond our current planning horizion / not well defined enough and put them into the Icebox for later. (Or deciding that we aren't going to do them at all and closing them if appropriate).

jdangerx replied:
I also wanted a bi-weekly process for pulling stuff out of Icebox! As you noted elsewhere, I stuffed that in mid-sprint checkin. But it doesn't have to be then. It's nice to have a week or so between doing this and the actual sprint starting, though, because fleshing issues out can take time and it's not always the best synchronous activity.

I think it'd be *nice* to have a higher cadence process for categorizing New stuff, though - something like "daily, a person looks at all the new things: categorize by team, and assess whether they go in the icebox or if they are urgent enough to shove into the sprint"

zaneselvans commented on docs/catalyst_101.md:
```diff
@@ -136,6 +136,159 @@ Once the existing members have voted to offer you membership, it’s up to you t
+
+#### Limit work-in-progress
+
+Goal: reduce overwhelm from "too many things going on at once" and identify bottlenecks in our process
```
On the flipside of too manything... sometimes I feel like there's nothing currently in a sprint that I have the necessary context to work on productively, and this can result in feeling like there's nothing I'm really allowed to work on. So either I have to sneak around and work on something I do know about without it being in the sprint, or work inefficiently on something I don't know enough about to actually be helpful.

jdangerx replied:
I think there are two ways we can take this:

* pull more stuff into the sprint when we run out of stuff within one person's context
* use this as a forcing function to share context between people

And I think this might be a bit of a case-by-case situation - i.e., is there other high priority stuff that one could knock out? (e.g. sloan stuff -> archiver)

Maybe this goes into "Help!" in the checkins - "Help! I don't know what to work on next."

zaneselvans commented on docs/catalyst_101.md:
```diff
@@ -136,6 +136,159 @@ Once the existing members have voted to offer you membership, it’s up to you t
+
+Here are future experiments:
+
+- breaking down tasks into smaller chunks - to be more flexible in how we divvy up the work, see incremental progress, and allow for faster review turnaround time
```
I've enjoyed using Task Lists to group smaller related tasks (or several issues) into bigger coherent chunks of work, and it sounds like this is the way GitHub is going for defining something like Epics. I like checking off the boxes and seeing the little progress bar go green. It also provides bidirectional tracks / tracked-by links on the issue pages. I think this kind of task/issue grouping would help make the process of gradually fleshing issues out as we understand more about what work is actually required, preserve some continuity of work streams across multiple sprints, give us a higher level view of what streams of work are in progress (by just looking at the issues tagged as Epics), and help keep a larger number of smaller issues from feeling cluttered or overwhelming.

jdangerx replied:
I think the projects integration with tasklists isn't *quite* there yet - or maybe it's just that we're not in the task list beta. There's a `tracks:` filter in GH projects but it doesn't let me filter to "tickets that are tracked by such-and-such epic" which would be, well, perfect.

zaneselvans replied:
The public version of Task Lists doesn't add explicit metadata to the  table / board views but it does ad tracks & tracked-by links within issues, and I to me that feels good enough for organizing grouped work (and it seems like the intention when the Task Lists beta goes public is that the metadata backing the feature would then become visible in the table / board views in a standard way).

zaneselvans commented on docs/catalyst_101.md:
```diff
@@ -136,6 +136,159 @@ Once the existing members have voted to offer you membership, it’s up to you t
+
+Activity:
+
+Use the mid sprint checkin as a parallel work block to pull items into Backlog - aiming for at least 2N items in Backlog, where N is the number of people on the team. All items in Backlog should have scope and a next step:
```
Ah okay I guess this where issues would go from New or Icebox to Backlog.

jdangerx commented on docs/catalyst_101.md:
```diff
@@ -136,6 +136,159 @@ Once the existing members have voted to offer you membership, it’s up to you t
+In the sprint review, instead of discussing status for each item in the sprint, prioritize the tickets that are in Backlog according to the following schema:
+
+- "Urgent" -> blocking work on high-priority tickets
+- "High" -> if we do not finish this during the next two weeks, we will not hit our quarterly goals
```
> My guess is that with these categories, we should be able to come up with enough Urgent and High priority items that nothing else ever gets worked on.

That's my guess too. At least, if we can get our planning/scoping heads screwed on straight (which is really hard!)

> I think there may need to be a parallel maintenance task track too -- things that need to be kept running alongside the new work. 

In terms of maintenance, I think we can frame a "keep things running" quarterly goal with metrics like:
* nightly builds run nightly
* datasette instance is up
* CI time < X minutes (ha!)
* XYZ datasets are archived successfully (some datasets might be more important than others)
* dependabot stuff is merged within X days (for which repos?)

In business jargon, we'd define Service Level Objectives and say "things that threaten to break our SLOs go into High priority."

bendnorman commented:
Hey @jdangerx! Do we still want to merge this in?

jdangerx commented:
Yes, but we probably need to revise this to match our process-as-practiced. I can put an hour or so into that and re-request review!

jdangerx commented:
I've updated this to reflect some changes:

* we moved to GitHub projects, hooray!
* we seem to have settled into a "set goals and use them to guide our sprints" process on Inframundo. Definitely things to improve there but it's pretty stable, so I just wrote down our process-as-practiced.
* I made it a little clearer that each team is its own island right now, but also we seem to all fulfill the same "put tickets on GH projects" contract.

bendnorman commented on docs/catalyst_101.md:
```diff
@@ -136,22 +136,55 @@ Once the existing members have voted to offer you membership, it’s up to you t
 
 ## Project Management
 
-### Tools
```
I think this tool information is still useful for new folks. What was your rational for removing it?

jdangerx replied:
I definitely wanted to remove the ZenHub language, and I didn't feel like these fit under "Project management" - I think it makes sense to pull them out into their own Communication + Admin Tools sections. It's useful to know what we intend to do with each of our many tools.

bendnorman reviewed (changes requested):
Looks good to me. Left a comment about removing tooling language.

bendnorman commented:
Also, I noticed 
```
Set up relevant accounts and familiarize yourself with our [project management tools](https://docs.google.com/document/d/1EkKfdW0Q1AIGexCflZMpw-4HuH1fJ-cdpi3ziancmlE/edit#heading=h.yeh6v1ee9ind) (reach out to Zane for assistance)
```
 Still links to the google doc.
